File: proxy.js
/**
 * Award Flight Search — Local Proxy Server (Node.js, zero dependencies)
 * Forwards browser requests to seats.aero with the Partner-Authorization header.
 *
 * Run:
 *   node proxy.js
 *
 * Then open index.html in your browser.
 */

const http  = require('http');
const https = require('https');
const url   = require('url');

const PORT = 5000;

const server = http.createServer((req, res) => {
  // Always add CORS headers so file:// pages can talk to us
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Headers', 'X-Api-Key, Content-Type');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if (req.url === '/health') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ status: 'ok', proxy: 'seats.aero', node: process.version }));
    return;
  }

  if (!req.url.startsWith('/proxy/search')) {
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Unknown route. Use /proxy/search' }));
    return;
  }

  const apiKey = req.headers['x-api-key'];
  if (!apiKey) {
    res.writeHead(400, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Missing X-Api-Key header.' }));
    return;
  }

  // Forward query string as-is to seats.aero
  const parsed   = url.parse(req.url, true);
  const qs       = new URLSearchParams(parsed.query).toString();
  const apiPath  = '/partnerapi/search' + (qs ? '?' + qs : '');

  const options = {
    hostname: 'seats.aero',
    path:     apiPath,
    method:   'GET',
    headers: {
      'Partner-Authorization': apiKey,
      'Accept':                'application/json',
      'User-Agent':            'AwardFlightSearch/1.0',
    },
  };

  console.log(`[proxy] GET /partnerapi/search ${qs.slice(0, 120)}`);

  const proxyReq = https.request(options, (proxyRes) => {
    res.writeHead(proxyRes.statusCode, {
      'Content-Type':                  'application/json',
      'Access-Control-Allow-Origin':   '*',
    });
    proxyRes.pipe(res);
  });

  proxyReq.setTimeout(45_000, () => {
    proxyReq.destroy();
    if (!res.headersSent) {
      res.writeHead(504, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'seats.aero request timed out (45s). Try a narrower date range.' }));
    }
  });

  proxyReq.on('error', (err) => {
    console.error('[proxy] error:', err.message);
    if (!res.headersSent) {
      res.writeHead(502, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'Proxy connection error: ' + err.message }));
    }
  });

  proxyReq.end();
});

server.listen(PORT, '127.0.0.1', () => {
  console.log('');
  console.log('  ✈  Award Flight Search Proxy  (Node.js ' + process.version + ')');
  console.log('  ' + '─'.repeat(42));
  console.log('  Listening:   http://localhost:5000');
  console.log('  Health:      http://localhost:5000/health');
  console.log('  ' + '─'.repeat(42));
  console.log('  Open index.html in your browser to start searching.');
  console.log('  Press Ctrl+C to stop.');
  console.log('');
});

File: proxy.py
"""
Seats.aero Award Flight Search — Local Proxy Server
Runs on http://localhost:5000 and forwards requests to the Seats.aero API.

Install dependencies:
    pip install flask flask-cors requests

Run:
    python proxy.py

Then open index.html in your browser.
"""

from flask import Flask, request, Response, jsonify
import requests
from flask_cors import CORS

app = Flask(__name__)
CORS(app)  # Allow requests from file:// and any localhost origin

SEATS_AERO_BASE = "https://seats.aero/partnerapi"


@app.route("/proxy/search")
def proxy_search():
    # Accept API key from custom header (sent by the browser JS)
    api_key = request.headers.get("X-Api-Key", "").strip()
    if not api_key:
        return jsonify({"error": "Missing API key. Send X-Api-Key header."}), 400

    # Forward all query params as-is to seats.aero
    params = dict(request.args)

    headers = {
        "Partner-Authorization": api_key,
        "Accept": "application/json",
        "User-Agent": "AwardFlightSearch/1.0",
    }

    try:
        resp = requests.get(
            f"{SEATS_AERO_BASE}/search",
            params=params,
            headers=headers,
            timeout=45,
        )
        # Stream the raw JSON back to the browser
        return Response(
            resp.content,
            status=resp.status_code,
            content_type="application/json",
        )
    except requests.exceptions.Timeout:
        return jsonify({"error": "Seats.aero request timed out (45s). Try a narrower date range."}), 504
    except requests.exceptions.RequestException as exc:
        return jsonify({"error": f"Proxy connection error: {exc}"}), 502


@app.route("/health")
def health():
    return jsonify({"status": "ok", "proxy": "seats.aero"})


if __name__ == "__main__":
    print()
    print("  ✈  Award Flight Search Proxy")
    print("  " + "─" * 36)
    print("  Listening on  http://localhost:5000")
    print("  Health check  http://localhost:5000/health")
    print("  " + "─" * 36)
    print("  Open index.html in your browser")
    print()
    app.run(debug=False, port=5000, host="127.0.0.1")
